### PR TITLE
Enable parallel resource creation

### PR DESCRIPTION
**Pull request by anuchandy**
[1]. CreatableImpl contains two public methods 'createRootResource' and 'createRootResourceAsync' those are simply wraps call to 'createResource' and 'createResourceAsync'. 'createResource' and 'createResourceAsync' are protected abstract methods that each resource implements, but this `protected` visibility won't really hide them since it's accessible via 'createRootResource' and 'createRootResourceAsync'. This PR removes unnecessary wrappers that achieves 2.

[2]. The above 'createRootResource' and 'createRootResourceAsync' are used in CreatableImpl::create & CreatableImpl::createAsync and the code is bit complex in these create methods since it is invoked in two contexts - preparer and non-preparer. This PR simplify the create methods. 

[3]. Addresses https://github.com/Azure/azure-sdk-for-java/issues/972 

[4]. As part of addressing 3, the task group now stores references to `ResourceCreator`

[5]. Renamed CreatableTaskGroup to CreatorTaskGroup as it hold ResourceCreator instances

[6]. Instead of picking one task at a time from the queue we pick all of them and run in parallel

[7]. Addressing https://github.com/Azure/azure-sdk-for-java/issues/980

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@anuchandy__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

**Comment by jianghaolu** [comment]
@azuresdkci test this please

**Comment by jianghaolu** [comment]
@azuresdkci add to whitelist

**Comment by azurecla** [comment]
Hi __@anuchandy__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

**Comment by anuchandy** [comment]
@martinsawicki @jianghaolu parallel resource creation work is done. Please review..

**Comment by anuchandy** [comment]
Here is a sample if you want to validate the concurrent async calls.
 
```java

package com.microsoft.azure.management.compute.samples;

import com.microsoft.azure.Azure;
import com.microsoft.azure.management.compute.KnownLinuxVirtualMachineImage;
import com.microsoft.azure.management.compute.VirtualMachine;
import com.microsoft.azure.management.compute.VirtualMachineSizeTypes;
import com.microsoft.azure.management.network.Network;
import com.microsoft.azure.management.network.NetworkInterface;
import com.microsoft.azure.management.network.PublicIpAddress;
import com.microsoft.azure.management.resources.ResourceGroup;
import com.microsoft.azure.management.resources.fluentcore.arm.Region;
import com.microsoft.azure.management.resources.fluentcore.utils.ResourceNamer;
import com.microsoft.azure.management.samples.Utils;
import com.microsoft.rest.ServiceCall;
import com.microsoft.rest.ServiceCallback;
import com.microsoft.rest.ServiceResponse;
import okhttp3.logging.HttpLoggingInterceptor;

import java.io.File;
import java.util.concurrent.CountDownLatch;

/**
 * Azure Compute sample for managing virtual machines.
 */
public final class ManageVirtualMachine2 {

    /**
     * Main entry point.
     * @param args the parameters
     */
    public static void main(String[] args) {
        final String rgName = Utils.createRandomName("rgCOMV");
        final String vnetName = ResourceNamer.randomResourceName("vnet", 24);
        final String nicName = ResourceNamer.randomResourceName("nic1", 24);
        final String nicName2 = ResourceNamer.randomResourceName("nic2", 24);
        final String pipDnsLabel = ResourceNamer.randomResourceName("pip1", 24);
        final String linuxVMName = Utils.createRandomName("lVM");
        final String userName = "tirekicker";
        final String password = "12NewPA$$w0rd!";
        final String pipName = ResourceNamer.randomResourceName("pip1", 20);

        try {

            //=============================================================
            // Authenticate

            final File credFile = new File(System.getenv("AZURE_AUTH_LOCATION"));

            Azure azure = Azure
                    .configure()
                    .withLogLevel(HttpLoggingInterceptor.Level.BODY)
                    .authenticate(credFile)
                    .withDefaultSubscription();

            // Print selected subscription
            System.out.println("Selected subscription: " + azure.subscriptionId());


            try {

                // Creatable RG
                ResourceGroup.DefinitionCreatable creatableRG = azure.resourceGroups()
                        .define(rgName)
                        .withRegion(Region.US_EAST);

                // Creatable Network
                Network.DefinitionStages.WithCreate creatableNetwork = azure.networks()
                        .define(vnetName)
                        .withRegion(Region.US_EAST)
                        .withNewResourceGroup(creatableRG)
                        .withAddressSpace("172.16.0.0/16")
                        .defineSubnet("Front-end")
                        .withAddressPrefix("172.16.1.0/24")
                        .attach()
                        .defineSubnet("Mid-tier")
                        .withAddressPrefix("172.16.2.0/24")
                        .attach()
                        .defineSubnet("Back-end")
                        .withAddressPrefix("172.16.3.0/24")
                        .attach();

                // Creatable Public Ip
                PublicIpAddress.DefinitionStages.WithCreate creatablePip = azure.publicIpAddresses()
                        .define(pipName)
                        .withRegion(Region.US_EAST)
                        .withNewResourceGroup(creatableRG)
                        .withLeafDomainLabel(pipDnsLabel);


                // Creatable Primary Nic
                NetworkInterface.DefinitionStages.WithCreate creatableNicPrimary = azure.networkInterfaces().define(nicName)
                        .withRegion(Region.US_EAST)
                        .withNewResourceGroup(creatableRG)
                        .withNewPrimaryNetwork(creatableNetwork)
                        .withPrimaryPrivateIpAddressDynamic()
                        .withNewPrimaryPublicIpAddress(creatablePip)
                        .withIpForwarding();

                // Creatable Secondary Nic
                NetworkInterface.DefinitionStages.WithCreate creatableNicSecondary = azure.networkInterfaces().define(nicName2)
                        .withRegion(Region.US_EAST)
                        .withNewResourceGroup(creatableRG)
                        .withNewPrimaryNetwork(creatableNetwork)
                        .withPrimaryPrivateIpAddressDynamic();

                // Create a Linux VM
                System.out.println("Creating a Linux VM in the network");

                final CountDownLatch latch = new CountDownLatch(1);

                ServiceCall serviceCall = azure.virtualMachines().define(linuxVMName)
                        .withRegion(Region.US_EAST)
                        .withExistingResourceGroup(rgName)
                        .withNewPrimaryNetworkInterface(creatableNicPrimary)
                        .withPopularLinuxImage(KnownLinuxVirtualMachineImage.UBUNTU_SERVER_16_04_LTS)
                        .withRootUserName(userName)
                        .withPassword(password)
                        .withSize(VirtualMachineSizeTypes.STANDARD_D3_V2)
                        .withNewSecondaryNetworkInterface(creatableNicSecondary)
                        .createAsync(new ServiceCallback<VirtualMachine>() {
                            @Override
                            public void failure(Throwable t) {
                                System.out.println(t.toString());
                                latch.countDown();
                            }

                            @Override
                            public void success(ServiceResponse<VirtualMachine> result) {
                                VirtualMachine linuxVM = result.getBody();
                                System.out.println("Created a Linux VM " + linuxVM.id());
                                Utils.print(linuxVM);
                                latch.countDown();
                            }
                        });

                // Wait for the async to complete
                latch.await();
            } catch (Exception f) {

                System.out.println(f.getMessage());
                f.printStackTrace();

            } finally {

                try {
                    System.out.println("Deleting Resource Group: " + rgName);
                    azure.resourceGroups().delete(rgName);
                    System.out.println("Deleted Resource Group: " + rgName);
                } catch (NullPointerException npe) {
                    System.out.println("Did not create any resources in Azure. No clean up is necessary");
                } catch (Exception g) {
                    g.printStackTrace();
                }

            }
        } catch (Exception e) {
            System.out.println(e.getMessage());
            e.printStackTrace();
        }
    }

    private ManageVirtualMachine2() {
    }
}
```

File: runtimes/azure-client-runtime/src/main/java/com/microsoft/azure/Graph.java [hunk 1]
```diff
@@ -13,6 +13,16 @@
 import java.util.Set;
 
 /**
+ * The edge types in a graph.
+ */
+enum GraphEdgeType {
```
Comment (jianghaolu) [inline runtimes/azure-client-runtime/src/main/java/com/microsoft/azure/Graph.java]: Add documentations here?

File: azure-mgmt-compute/src/main/java/com/microsoft/azure/management/compute/implementation/AvailabilitySetImpl.java [hunk 2]
```diff
@@ -101,28 +101,32 @@ public ServiceCall applyAsync(ServiceCallback<AvailabilitySet> callback) {
         return this.createAsync(callback);
     }
 
+    // CreatorTaskGroup.ResourceCreator implementation
+
     @Override
-    protected void createResource() throws Exception {
+    public Resource createResource() throws Exception {
```
Comment (jianghaolu) [inline azure-mgmt-compute/src/main/java/com/microsoft/azure/management/compute/implementation/AvailabilitySetImpl.java]: Makes sense to change `void` to `Resource` but why public?

Comment (anuchandy) [inline azure-mgmt-compute/src/main/java/com/microsoft/azure/management/compute/implementation/AvailabilitySetImpl.java]: we are trying to simplfy the logic we had earleier by removing two PUBLIC methods `createRootResource` and `createRootResourceAsync` exposed in every resource (via CreatableImpl inheritance). These two PUBLIC methods wraps calls to PROTECTED `createResource` and `createResourceAsync` methods. Removing createRoot* methods caused making the methods its wraps to public.

File: runtimes/azure-client-runtime/src/main/java/com/microsoft/azure/Graph.java (hunk 1 above)
Comment (anuchandy) [inline runtimes/azure-client-runtime/src/main/java/com/microsoft/azure/Graph.java]: Do you mean documentation for each entries in this enum?

File: azure-mgmt-compute/src/main/java/com/microsoft/azure/management/compute/implementation/AvailabilitySetImpl.java (hunk 2 above)
Comment (anuchandy) [inline azure-mgmt-compute/src/main/java/com/microsoft/azure/management/compute/implementation/AvailabilitySetImpl.java]: Btw I tried to explain the changes in this PR in the PR description, i can expand the description if anything is unclear.

Comment (jianghaolu) [inline azure-mgmt-compute/src/main/java/com/microsoft/azure/management/compute/implementation/AvailabilitySetImpl.java]: Got it

File: runtimes/azure-client-runtime/src/main/java/com/microsoft/azure/Graph.java (hunk 1 above)
Comment (jianghaolu) [inline runtimes/azure-client-runtime/src/main/java/com/microsoft/azure/Graph.java]: yeah that's a requirement for public enums but it's also helpful to understand what each type of edge means here?

**Comment by jianghaolu** [comment]
A couple comments, otherwise LGTM

File: runtimes/azure-client-runtime/src/main/java/com/microsoft/azure/Graph.java (hunk 1 above)
Comment (anuchandy) [inline runtimes/azure-client-runtime/src/main/java/com/microsoft/azure/Graph.java]: right, added them.

**Comment by anuchandy** [comment]
@azuresdkci test this please

**Comment by azurecla** [comment]
Hi __@anuchandy__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

**Pull request merged**